// primeng/esm5/dynamicdialog/dynamicdialogcontent.js
import { Directive, ViewContainerRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var DynamicDialogContent = /** @class */ (function () {
    function DynamicDialogContent(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
    }
    DynamicDialogContent.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    DynamicDialogContent.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
DynamicDialogContent.ɵfac = function DynamicDialogContent_Factory(t) { return new (t || DynamicDialogContent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
DynamicDialogContent.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: DynamicDialogContent, selectors: [["", "pDynamicDialogContent", ""]] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DynamicDialogContent, [{
        type: Directive,
        args: [{
                selector: '[pDynamicDialogContent]'
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }]; }, null); })();
    return DynamicDialogContent;
}());
export { DynamicDialogContent };
